Cython/Includes/cpython/datetime.pxd

```diff
@@ -1,27 +1,106 @@
 
 cdef extern from "Python.h":
     ctypedef struct PyTypeObject:
         pass
 
 cdef extern from "datetime.h":
     """
-    #if PY_MAJOR_VERSION < 3 && !defined(PyDateTime_DELTA_GET_DAYS)
-    #define PyDateTime_DELTA_GET_DAYS(o)         (((PyDateTime_Delta*)o)->days)
+    /* Backport for Python < 3.3 */
+    #if PY_VERSION_HEX < 0x030100a1
+        #ifndef PyDateTime_DELTA_GET_DAYS
+            #define PyDateTime_DELTA_GET_DAYS(o) (((PyDateTime_Delta*)o)->days)
+        #endif
+        #ifndef PyDateTime_DELTA_GET_SECONDS
+            #define PyDateTime_DELTA_GET_SECONDS(o) (((PyDateTime_Delta*)o)->seconds)
+        #endif
+        #ifndef PyDateTime_DELTA_GET_MICROSECONDS
+            #define PyDateTime_DELTA_GET_MICROSECONDS(o) (((PyDateTime_Delta*)o)->microseconds)
+        #endif
     #endif
-    #if PY_MAJOR_VERSION < 3 && !defined(PyDateTime_DELTA_GET_SECONDS)
-    #define PyDateTime_DELTA_GET_SECONDS(o)      (((PyDateTime_Delta*)o)->seconds)
+
+    /* Backport for Python < 3.6 */
+    #if PY_VERSION_HEX < 0x030600a4
+        #ifndef PyDateTime_TIME_GET_FOLD
+            #define PyDateTime_TIME_GET_FOLD(o) 0
+        #endif
+        #ifndef PyDateTime_DATE_GET_FOLD
+            #define PyDateTime_DATE_GET_FOLD(o) 0
+        #endif
+    #endif
+
+    /* Backport for Python < 3.6 */
+    #if PY_VERSION_HEX < 0x030600a4
+        #ifndef __Pyx_DateTime_DateTimeWithFold
+            #define __Pyx_DateTime_DateTimeWithFold(year, month, day, hour, minute, second, microsecond, tz, fold) \
+                PyDateTimeAPI->DateTime_FromDateAndTime(year, month, day, hour, minute, second, \
+                    microsecond, tz, PyDateTimeAPI->DateTimeType)
+        #endif
+        #ifndef __Pyx_DateTime_TimeWithFold
+            #define __Pyx_DateTime_TimeWithFold(hour, minute, second, usecond, tz, fold) \
+                PyDateTimeAPI->Time_FromTime(hour, minute, second, usecond, tz, PyDateTimeAPI->TimeType)
+        #endif
+    #else /* For Python 3.6+ so that we can pass tz */
+        #ifndef __Pyx_DateTime_DateTimeWithFold
+            #define PyDateTime_FromDateAndTimeAndFold(hour, minute, second, usecond, tz, fold) \
+                PyDateTimeAPI->DateTime_FromDateAndTimeAndFold(year, month, day, hour, minute, second, \
+                    microsecond, tz, fold, PyDateTimeAPI->DateTimeType)
+        #endif
+        #ifndef __Pyx_DateTime_TimeWithFold
+            #define __Pyx_DateTime_TimeWithFold(hour, minute, second, usecond, tz, fold) \
+                PyDateTimeAPI->Time_FromTimeAndFold(hour, minute, second, usecond, tz, fold, PyDateTimeAPI->TimeType)
+        #endif
     #endif
-    #if PY_MAJOR_VERSION < 3 && !defined(PyDateTime_DELTA_GET_MICROSECONDS)
-    #define PyDateTime_DELTA_GET_MICROSECONDS(o)   (((PyDateTime_Delta*)o)->microseconds)
+
+    /* Backport for Python < 3.7 */
+    #if PY_VERSION_HEX < 0x030700b1
+        #ifndef PyDateTime_TimeZone_UTC
+            #define PyDateTime_TimeZone_UTC NULL
```

Comment:
Since these are incomplete/incorrect backports (e.g. they return `NULL` but don't set an exception), I'd also use `__Pyx_` macros here.

Reply:
They do in our inline functions, but yes not on the C level.